### PR TITLE
Add basic CMSIS core headers

### LINKED ISSUE #21: Add CMSIS-Core headers
Please add standard CMSIS-Core headers and register definitions. Ideally these would replace the bespoke headers to avoid fracturing of the ecosystem.

The common CMSIS header files are [here](https://github.com/ARM-software/CMSIS_5/tree/develop/CMSIS/Core/Include). It's not uncommon to include the entire CMSIS_5 repo as a submodule, and it's just as common to copy the headers into your SDK.

[CMSIS-Core documentation](https://arm-software.github.io/CMSIS_5/Core/html/index.html)

CMSIS-standard register definitions can be generated with the [SVDConv utility](https://arm-software.github.io/CMSIS_5/SVD/html/svd_SVDConv_pg.html).

### PR DESCRIPTION
**Pull request by kilograham**
Adds a new cmsis_core INTERFACE library, which provides the basic CMSIS macros/definitions, and RP2040.h and system_RP2040.h
By default the isr_ methods in the vector table are renamed to their CMSIS versions, but this can be overridden by defining PICO_CMSIS_RENAME_EXCEPTIONS=0

fixes #21

### CODE REVIEW

**Comment by kilograham** [comment]
This depends on #382 for the exception/irq handler renaming

File: src/rp2_common/cmsis/include/cmsis/rename_exceptions.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2020 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef _CMSIS_RENAME_EXCEPTIONS_H
+#define _CMSIS_RENAME_EXCEPTIONS_H
+
+// PICO_CONFIG: PICO_CMSIS_RENAME_EXCEPTIONS, Whether to rename SDK exceptions such as isr_nmi to their CMSIS equivalent i.e. NMI_Handler, type=bool, default=0, group=cmsis_core
```
Comment (lurch) [inline src/rp2_common/cmsis/include/cmsis/rename_exceptions.h:10]: This line says `default=0`, but line 14 seems to set the default to 1? :confused:

**Comment by lurch** [comment]
FYI I didn't review this previously, because it was marked as a Draft PR.

File: src/rp2_common/cmsis/include/cmsis/rename_exceptions.h (hunk 1 above)
Comment (kilograham) [reply src/rp2_common/cmsis/include/cmsis/rename_exceptions.h:10]: please open a PR

**Comment by kilograham** [comment]
> FYI I didn't review this previously, because it was marked as a Draft PR.

ah, makes sense!

File: src/rp2_common/cmsis/include/cmsis/rename_exceptions.h (hunk 1 above)
Comment (lurch) [reply src/rp2_common/cmsis/include/cmsis/rename_exceptions.h:10]: What's the intended behaviour - change the PICO_CONFIG default to 1, or change the actual #define to 0 ?

Comment (kilograham) [reply src/rp2_common/cmsis/include/cmsis/rename_exceptions.h:10]: yeah the #define is correct i.e. 1

**Comment by lurch** [comment]
ping @ghollingworth for queries about the mixed-licensing.

**Review by liamfraser — COMMENTED** [review]
LGTM

**Comment by ghollingworth** [comment]
Graham,

We should add the CMSIS license to this LICENSE file and just list the path of files that are covered by it.

Gordon

**Comment by kilograham** [comment]
@ghollingworth you are a day late and a dollar short. this is already merged/released. Also

> We should add the CMSIS license to this LICENSE file and just list the path of files that are covered by it.

What is _this_ license? list he files where?

might be simplest to submit a PR.

**Comment by lurch** [comment]
I suspect Gordon might have been thinking something along the lines of https://github.com/raspberrypi/firmware/blob/master/README.md ? :shrug:

**Comment by ghollingworth** [comment]
No I was thinking along the lines of the debian copyright format:

https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/

This allows something like the following:

```
Files: *
Copyright: 1975-2010 Ulla Upstream
License: GPL-2+

Files: debian/*
Copyright: 2010 Daniela Debianizer
License: GPL-2+

Files: debian/patches/fancy-feature
Copyright: 2010 Daniela Debianizer
License: GPL-3+

Files: */*.1
Copyright: 2010 Manuela Manpager
License: GPL-2+
```

**Pull request merged**